ikarus/finiteelements/mechanics/materials/hyperelastic/interface.hh

```diff
@@ -0,0 +1,348 @@
+   */
+  template <typename STO>
+  auto rebind() const {
+    auto reboundDEV = dev_.template rebind<STO>();
+    auto reboundVOL = vol_.template rebind<STO>();
+    return Hyperelastic<decltype(reboundDEV), decltype(reboundVOL)>(reboundDEV, reboundVOL);
+  }
+
+private:
+  DEV dev_;
+  VOL vol_;
+
+  inline static constexpr auto dimensionRange() { return Dune::range(dim); }
+
+  template <typename ST>
+  Eigen::Matrix<ST, 3, 3> transformDeviatoricStresses(const Eigen::Matrix<ST, 3, 1>& principalStress,
+                                                      const Eigen::Matrix<ST, 3, 3>& N) const {
+    return (N * principalStress.asDiagonal() * N.transpose()).eval();
+  }
+
+  template <typename ST>
+  Eigen::Matrix<ST, 3, 3> transformVolumetricStresses(ST Uprime, const auto& C, ST J) const {
+    return J * Uprime * C.inverse();
+  }
+
+  template <typename ST>
+  auto transformDeviatoricTangentModuli(const Eigen::TensorFixedSize<ST, Eigen::Sizes<3, 3, 3, 3>>& L,
+                                        const Eigen::Matrix<ST, 3, 3>& N) const {
+    Eigen::TensorFixedSize<ST, Eigen::Sizes<3, 3, 3, 3>> moduli{};
+    moduli.setZero();
+
+    for (const auto i : dimensionRange())
+      for (const auto k : dimensionRange()) {
+        // First term: L[i, i, k, k] * ((N[i] ⊗ N[i]) ⊗ (N[k] ⊗ N[k]))
+        auto NiNi = dyadic(N.col(i).eval(), N.col(i).eval());
+        auto NkNk = dyadic(N.col(k).eval(), N.col(k).eval());
+
+        moduli += L(i, i, k, k) * dyadic(NiNi, NkNk);
+
+        // Second term (only if i != k): L[i, k, i, k] * (N[i] ⊗ N[k] ⊗ (N[i] ⊗ N[k] + N[k] ⊗ N[i]))
+        if (i != k) {
+          auto NiNk = dyadic(N.col(i).eval(), N.col(k).eval());
+          auto NkNi = dyadic(N.col(k).eval(), N.col(i).eval());
+
+          moduli += L(i, k, i, k) * dyadic(NiNk, NiNk + NkNi);
+        }
+      }
+
+    return moduli;
+  }
+
+  template <typename ST>
+  auto transformVolumetricTangentModuli(const ST& Uprime, const ST& Uprimeprime, const auto& C, ST J) const {
+    const auto invC  = C.inverse().eval();
+    const auto CTinv = tensorView(invC, std::array<Eigen::Index, 2>({3, 3}));
+
+    Eigen::TensorFixedSize<ST, Eigen::Sizes<3, 3, 3, 3>> moduli =
+        (J * ((Uprime + J * Uprimeprime) * dyadic(CTinv, CTinv) -
+              (2 * Uprime * symTwoSlots(fourthOrderIKJL(invC, invC), {2, 3}))))
+            .eval();
+
+    return moduli;
+  }
+
+  template <typename Derived>
+  auto principalStretches(const Eigen::MatrixBase<Derived>& Craw, int options = Eigen::ComputeEigenvectors) const {
+    auto C = Impl::maybeFromVoigt(Craw);
+    return Impl::principalStretches<typename Derived::Scalar>(C, options);
+  }
+
+  template <typename ST>
+  auto detF(const Eigen::Array<ST, 3, 1>& lambda) const -> ST {
+    const auto detC = Impl::determinantFromPrincipalValues<ST>(lambda);
+    Impl::checkPositiveDet(detC);
+    return detC;
+  }
+
+  // Unpack the derivatives from the result of an @ref eval call into an array.
```

Comment:
This comment does not reflect what this function does. Furthermore it should not inside the class. This is actually not needed at all. My godbolt code was not mature! It only was a prrof of concept